stripe-ui-core/src/main/java/com/stripe/android/uicore/elements/compat/TextFieldLayout.kt

```diff
@@ -0,0 +1,483 @@
+        // top with padding
+        val startPosition = if (singleLine) {
+            Alignment.CenterVertically.align(it.height, height)
+        } else {
+            // even though the padding is defined by developer, it only affects text field when animation
+            // progress == 1, which is when text field is focused or non-empty input text. The start position of the
+            // label is always 16.dp.
+            (TextFieldPadding.value * density).roundToInt()
+        }
+        val distance = startPosition - labelEndPosition
+        val positionY = startPosition - (distance * animationProgress).roundToInt()
+        it.placeRelative(widthOrZero(leadingPlaceable), positionY)
+    }
+    textfieldPlaceable.placeRelative(widthOrZero(leadingPlaceable), textPosition)
+    placeholderPlaceable?.placeRelative(widthOrZero(leadingPlaceable), textPosition)
+}
+
+/**
+ * Places the provided text field and placeholder in [TextField] when there is no label. When
+ * there is a label, [placeWithLabel] is used
+ */
+private fun Placeable.PlacementScope.placeWithoutLabel(
+    width: Int,
+    height: Int,
+    textPlaceable: Placeable,
+    placeholderPlaceable: Placeable?,
+    leadingPlaceable: Placeable?,
+    trailingPlaceable: Placeable?,
+    singleLine: Boolean,
+    density: Float,
+    paddingValues: PaddingValues
+) {
+    val topPadding = (paddingValues.calculateTopPadding().value * density).roundToInt()
+
+    leadingPlaceable?.placeRelative(
+        0,
+        Alignment.CenterVertically.align(leadingPlaceable.height, height)
+    )
+    trailingPlaceable?.placeRelative(
+        width - trailingPlaceable.width,
+        Alignment.CenterVertically.align(trailingPlaceable.height, height)
+    )
+
+    // Single line text field without label places its input center vertically. Multiline text
+    // field without label places its input at the top with padding
+    val textVerticalPosition = if (singleLine) {
+        Alignment.CenterVertically.align(textPlaceable.height, height)
+    } else {
+        topPadding
+    }
+    textPlaceable.placeRelative(
+        widthOrZero(leadingPlaceable),
+        textVerticalPosition
+    )
+
+    // placeholder is placed similar to the text input above
+    placeholderPlaceable?.let {
+        val placeholderVerticalPosition = if (singleLine) {
+            Alignment.CenterVertically.align(placeholderPlaceable.height, height)
+        } else {
+            topPadding
+        }
+        it.placeRelative(
+            widthOrZero(leadingPlaceable),
+            placeholderVerticalPosition
+        )
+    }
+}
+
+private val IntrinsicMeasurable.layoutId: Any?
+    get() = (parentData as? LayoutIdParentData)?.layoutId
+
+private fun widthOrZero(placeable: Placeable?) = placeable?.width ?: 0
+private fun heightOrZero(placeable: Placeable?) = placeable?.height ?: 0
+
+private val TextFieldTopPadding = 4.dp
```

Comment:
I'm not a fan of effectively needing to import the entire execution pathway of `TextField` to `TextFieldLayout` to change this value from `2.dp` back to `4.dp` but its the only way I've thought of to maintain compatibility. There was a potentially workaround in modifying the bottom padding to account for it by the code in `calculateHeight` doesn't allow for unfortunately since it multiples the padding values by the density
```
private fun calculateHeight(
    textFieldHeight: Int,
    hasLabel: Boolean,
    labelBaseline: Int,
    leadingHeight: Int,
    trailingHeight: Int,
    placeholderHeight: Int,
    constraints: Constraints,
    density: Float,
    paddingValues: PaddingValues
): Int {
    val paddingToLabel = TextFieldTopPadding.value * density
    val topPaddingValue = paddingValues.calculateTopPadding().value * density
    val bottomPaddingValue = paddingValues.calculateBottomPadding().value * density

    val inputFieldHeight = max(textFieldHeight, placeholderHeight)
    val middleSectionHeight = if (hasLabel) {
        labelBaseline + paddingToLabel + inputFieldHeight + bottomPaddingValue
    } else {
        topPaddingValue + inputFieldHeight + bottomPaddingValue
    }
    return maxOf(
        middleSectionHeight.roundToInt(),
        max(leadingHeight, trailingHeight),
        constraints.minHeight
    )
}
```